Overhaul truncate and snapshot

bhalevy opened the pull request:
This series is aimed at fixing #11132.

To get there, the series untangles the functions that currently depend on the the cross-shard coordination in table::snapshot,
namely database::truncate and consequently database::drop_column_family.

database::get_table_on_all_shards is added here as a helper to get a foreign shared ptr of the the table shard from all shards,
and it is later used by multiple functions to truncate and then take a snapshot of the sharded table.

database::truncate_table_on_all_shards is defined to orchestrate the truncate process end-to-end, flushing or clearing all table shards before taking a snapshot if needed, using the newly defined table::snapshot_on_all_shards, and by that leaving only the discard_sstables job to the per-shard database::truncate function.

The latter, snapshot_on_all_shards, orchestrates the snapshot process on all shards - getting rid of the per-shard table::snapshot function (after refactoring take_snapshot and finalize_snapshot out of it), and the associated dreaded data structures: snapshot_manager and pending_snapshots.

Fixes #11132.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1525/

bhalevy commented:
@nyh since you reviewed changes to this code path recently, can you please review this one too?

denesb commented on replica/database.hh:
```diff
@@ -857,7 +857,9 @@ public:
     db::replay_position set_low_replay_position_mark();
 
 private:
-    future<std::unordered_set<sstring>> take_snapshot(database& db, sstring jsondir);
```
Why is copying sstring instances across shard fishy? They might result in an unnecessary copy, but I don't see anything fishy here. Also, this code doesn't seem to care for performance much, prioritizing ease of reading/writing, which is not helped by this foreign pointer.

avikivity replied:
agree

bhalevy replied:
Currently the strings are allocated and freed on the calling coroutine frame
and we submit_to the "coordinator" shard to copy them across shard.

But the refactoring done here changes that, where the coordinator shard will submit_to all shards to perform `take_snapshot`, returning the `std::unordered_set<sstring>` in the future result.

We cannot free that on the coordinator shard anymore, so we need to wrap it in a foreign ptr so it will get freed on the shard that allocated it.

Looks like I should improve the explanation about the motivation for this patch...

denesb commented on replica/database.hh:
```diff
@@ -861,7 +861,8 @@ private:
     future<snapshot_file_set> take_snapshot(database& db, sstring jsondir);
-    future<> snapshot(database& db, sstring name);
+    // FIXME: delete _unused_snapshot method
+    future<> _unused_snapshot(database& db, sstring name);
```
:)

denesb commented on replica/table.cc:
```diff
@@ -1262,6 +1262,7 @@ table::seal_snapshot(sstring jsondir, std::vector<snapshot_file_set> file_sets)
             ss << ", ";
         }
         ss << "\"" << rf << "\"";
+        co_await coroutine::maybe_yield();
```
unrelated change?

bhalevy replied:
it's orthogonal indeed.
I can separate it into a patch of its own.

denesb reviewed (commented):
I cannot claim to have completely followed all these changes, but the overall plan looks good to me.
Are the new variants of snapshot and truncate re-entrant?
How are failures handled compared to the previous model?

bhalevy commented on replica/database.cc:
```diff
@@ -1032,11 +1025,28 @@ future<> database::drop_column_family(const sstring& ks_name, const sstring& cf_
     f.get(); // re-throw exception from truncate() if any
 }
 
+future<std::vector<foreign_ptr<lw_shared_ptr<table>>>> database::get_table_on_all_shards(sharded<database>& sharded_db, utils::UUID uuid) {
```
We want to extract the `lw_shared_ptr<table>` instances before they get dropped by `drop_table_on_all_shards`.
By the time we get to snapshot and truncate the table on each shard the instances are not listed anymore.

raphaelsc replied:
> We want to extract the `lw_shared_ptr<table>` instances before they get dropped by `drop_table_on_all_shards`. By the time we get to snapshot and truncate the table on each shard the instances are not listed anymore.

But if we lookup the table before drop_column_family() is called (as today), doesn't it work?

dropping a table is always the last step.

raphaelsc replied:
Now I understand it, it's a preexisting requirement. That's why drop procedure has to keep a shared ptr to table.

bhalevy replied:
right.  We do look it up on each shard using either uuid or ks/cf names, and keep the shared pointer to be used later on, after the table shard-instances get unlisted.

As a follow up, I think we can change the design so to have sharded<keyspace> and shard<table> instances (deriving form peering_sharded_service), so we can get to all shard instances using the local instance, since neither are really independent. I.e. you can have a keyspace or a table instantiated not on all shards.

bhalevy commented on replica/database.cc:
```diff
@@ -1046,8 +1043,13 @@ future<> database::drop_table_on_all_shards(sharded<database>& sharded_db, sstri
+    co_await smp::invoke_on_all([&] {
+        return table_shards[this_shard_id()]->stop();
     });
+    f.get(); // re-throw exception from truncate() if any
```
Nope, although an error will be returned, the table is dropped at this point.

raphaelsc commented on service/storage_proxy.cc:
```diff
@@ -5436,12 +5436,8 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
-        });
+    slogger.info("Truncting {}.{}", ksname, cfname);
+    return replica::database::truncate_table_on_all_shards(_db, ksname, cfname, [ts = db_clock::now()] { return make_ready_future<db_clock::time_point>(ts); });
 }
```
interesting, previously we could have different shards picking different timestamps, as the callback calculated a timestamp when called, not upfront.

example: shard s1 picks T1. shard s2 stalls, and picks a different timestamp T2, where T2 > T1.

I see the timestamp is used for creation of snapshot directory in database::truncate():

```
    db_clock::time_point truncated_at = co_await tsf();

    if (auto_snapshot) {
        auto name = format("{:d}-{}", truncated_at.time_since_epoch().count(), cf.schema()->cf_name());
        co_await cf.snapshot(*this, name);
    }
```

So previously, it could happen that snapshot could pick different directories in different shards, right?

You're making the code more robust here, and you're actually fixing a problem and an issue should be opened?!

raphaelsc replied:
Turns out joinpoint provides the correct semantics. sorry for the noise.

bhalevy replied:
yes, but the caller could theoretically pass a timestamp_func that will return the current time on each shard, but that won't work, as you noticed.

joinpoint is a subtle and unvisible way of synchronizing between the shard via this callback.
I think it was a bad idea to use that here.

raphaelsc commented on service/storage_proxy.cc:
```diff
@@ -5436,12 +5436,8 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
-                return sp._db.local().truncate(ksname, cfname, [&tsf] { return tsf.value(); });
-            });
-        });
+    slogger.info("Truncting {}.{}", ksname, cfname);
```
Typo: Truncting -> Truncating

raphaelsc commented on replica/database.cc:
```diff
@@ -1025,7 +1025,7 @@ future<> database::drop_column_family(const sstring& ks_name, const sstring& cf_
     co_await cf->await_pending_ops();
-    co_await _querier_cache.evict_all_for_table(cf->schema()->id());
+    co_await _querier_cache.evict_all_for_table(uuid);
     auto f = co_await coroutine::as_future(truncate(*cf, std::move(tsf), snapshot));
```
The schema UUID is different than the UUID used to globally identify a ks.cf pair (which can be found through database::find_uuid()). I think it's worth adding a comment at least.

avikivity replied:
And this is a very subtle point. We need to use strong types and only convert from plain uuid during communications.

raphaelsc replied:
> That means this change is wrong, no?

yes. It's very easy to confuse table UUID and schema UUID, and using them in incorrect place means a bug, as they represent different things.

+1 for using strong types for identifiers.

e.g.:

replica::table::uuid_type

schema::uuid_type.

raphaelsc replied:
But that's a change for another series.

bhalevy replied:
> The schema UUID is different than the UUID used to globally identify a ks.cf pair (which can be found through database::find_uuid()). I think it's worth adding a comment at least.

They are the same.
See https://github.com/scylladb/scylladb/blob/768a5c8b5a1eef0086eb9186924df7aee01d1b57/replica/database.cc#L959-L970

(Unlike the schema version() that returns the `table_schema_version` as UUID too)

raphaelsc replied:
hmm schema->id() refers to the immutable table UUID, not schema version uuid. that's a bit confusing. perhaps id() should be renamed to table_id(). and switch to strong types for them as suggested by Avi.

bhalevy replied:
Will so as a followup

bhalevy replied:
#11207

raphaelsc replied:
thanks!

raphaelsc commented on replica/table.cc:
```diff
@@ -1379,10 +1379,12 @@ future<std::unordered_set<sstring>> table::take_snapshot(database& db, sstring j
         auto f = sst->get_filename();
         auto rf = f.substr(sst->get_dir().size() + 1);
         table_names.insert(std::move(rf));
+        co_await coroutine::maybe_yield();
```
if we consider each iteration takes 1us, then we indeed have a potential for few ms stalls here.

raphaelsc replied:
thousand sstable threshold can be easily hit with off-strategy compaction.

bhalevy replied:
right

avikivity replied:
I don't think an iteration can take a microsecond here, but okay.

raphaelsc commented on replica/database.cc:
```diff
@@ -1013,31 +1013,46 @@ void database::remove(const table& cf) noexcept {
+}
+
+future<std::vector<foreign_ptr<lw_shared_ptr<table>>>> database::get_table_on_all_shards(sharded<database>& sharded_db, utils::UUID uuid) {
+    std::vector<foreign_ptr<lw_shared_ptr<table>>> table_shards;
```
perhaps we can make foreign_ptr<lw_shared_ptr<table>>> easier to read with:

using foreign_table_ptr = make foreign_ptr<lw_shared_ptr<table>>>

bhalevy replied:
yeah.  BTW, I once had a proposal to define short names for `foreign_unique_ptr`, `foreign_lw_shared_ptr`, `foreign_shared_ptr`, etc.  but it didn't go through.

raphaelsc commented on replica/database.hh:
```diff
@@ -857,10 +857,16 @@ public:
-    friend class database;
 public:
+    static future<> snapshot_on_all_shards(sharded<database>& sharded_db, const std::vector<foreign_ptr<lw_shared_ptr<table>>>& table_shards, sstring name);
+
```
snapshot_on_all_shards() belongs better to database, but it seems you placed it here to access snapshot functions in table's private namespace.

bhalevy replied:
Right. There's a corresponding `replica::database::snapshot_table_on_all_shards` entry point.
Maybe it'd be better to declare it here private/protected and keep `friend class database`, but I don't like the database messing up with table's privates...

raphaelsc commented on replica/database.cc:
```diff
@@ -1013,31 +1013,46 @@ void database::remove(const table& cf) noexcept {
+}
+
+future<std::vector<foreign_ptr<lw_shared_ptr<table>>>> database::get_table_on_all_shards(sharded<database>& sharded_db, utils::UUID uuid) {
+    std::vector<foreign_ptr<lw_shared_ptr<table>>> table_shards;
```
By the way, we could also abstract access to local table, with something like this:

```
class global_table_ptr {
    std::vector<foreign_ptr<lw_shared_ptr<table>>> _table_per_shard;
public:
    static global_table_ptr make_global_table_ptr(sharded<database>& sharded_db, utils::UUID uuid);
    
    table& operator->() const {
        return *_table_per_shard[this_shard_id()];
    }
};
    
```

maybe it's overkill. I like the approach you used too.

bhalevy replied:
as I said above, I think that the longer term approach should be in the direction of `sharded<keyspace>` and `sharded<table>` objects.  Not sure how to stitch them to the database shards (maybe define a table::impl and have table contain a lw_shared_ptr<table::impl>).

raphaelsc replied:
If database itself is sharded, then theoretically table and keyspace are just members of database and don't have to be sharded. Cross-shard operations requiriring a coordinator should go through database which is itself sharded.

raphaelsc replied:
But with an eye on tablets, the idea is to decouple table objects from database objects, which has a single instance per shard and won't move.

raphaelsc commented on replica/table.cc:
```diff
@@ -1300,130 +1272,118 @@ seal_snapshot(sstring jsondir) {
+
+        co_await coroutine::parallel_for_each(boost::irange(0u, smp::count), [&] (unsigned shard) -> future<> {
+            file_sets.emplace_back(co_await smp::submit_to(shard, [&] {
+                return table_shards[this_shard_id()]->take_snapshot(sharded_db.local(), jsondir);
```
I think vector's emplace back is not thread safe.

Better to place snapshot_file_set in the idx of the originating shard.

raphaelsc replied:
please disregard my comment. we're doing the insertion via the coordinator.

raphaelsc reviewed (commented):
the direction looks very good. just had a few comments.

bhalevy commented on replica/database.cc:
```diff
@@ -2344,15 +2344,20 @@ future<> database::flush_keyspace_on_all_shards(sharded<database>& sharded_db, s
+}
+
+future<> database::snapshot_tables_on_all_shards(sharded<database>& sharded_db, std::string_view ks_name, std::vector<sstring> table_names, sstring tag, bool skip_flush) {
+    co_await coroutine::parallel_for_each(table_names, [&, ks_name, tag = std::move(tag), skip_flush] (const auto& table_name) -> future<> {
```
I'll see if legacy parallel_for_each can be used here without need for extra allocation.

avikivity replied:
I think it's okay to use coroutines for administrative code, since it's likely to gain more code and emphasis should be on safety and readability, not performance.

bhalevy commented on replica/database.cc:
```diff
@@ -2395,13 +2403,6 @@ future<> database::truncate(column_family& cf, timestamp_func tsf, bool with_sna
-    // Also, the replay_position returned by discard_sstables() may refer to old commit log domain.
-    if (cf.schema()->ks_name() == db::schema_tables::NAME) {
-        throw std::runtime_error(format("Truncating of {}.{} is not allowed.", cf.schema()->ks_name(), cf.schema()->cf_name()));
-    }
```
It becomes private in this series.

avikivity commented on replica/database.cc:
```diff
@@ -1013,16 +1013,9 @@ void database::remove(const table& cf) noexcept {
-    table& cf = *cfptr;
+future<> database::drop_column_family(table& cf, timestamp_func tsf, bool snapshot) {
+    auto uuid = cf.schema()->id();
+    drop_repair_history_map_for_table(uuid);
```
This should be the last step, since it can race with a compaction completing just now.

Anyway, no need to touch this for this series.

bhalevy replied:
Makes sense

avikivity commented on replica/database.cc:
```diff
@@ -1013,16 +1013,13 @@ void database::remove(const table& cf) noexcept {
 }
 
-future<> database::drop_column_family(table& cf, timestamp_func tsf, bool snapshot) {
+future<> database::drop_column_family(table& cf) {
```
Maybe it's now better named detach_column_family(), since it just disassociates the table from the APIs, bit it keeps existing in the background.

avikivity replied:
(and it's just a small part of the drop procedure)

avikivity replied:
For bonus points return the foreign_ptr here. So this function detaches ownership from database and gives it to the caller. But that's just cosmetic, ignore if it causes problems.

avikivity commented on replica/database.cc:
```diff
@@ -1049,7 +1049,10 @@ future<> database::drop_table_on_all_shards(sharded<database>& sharded_db, sstri
+    if (f.failed()) {
+        co_await coroutine::return_exception_ptr(f.get_exception());
+    }
     co_await sstables::remove_table_directory_if_has_no_snapshots(table_dir);
```
I think it's just noise. Yes, we avoid exception rethrowing. But is it important here?

bhalevy replied:
It's not important here.  I just got used to it.
I can drop this patch. It's immaterial indeed.

avikivity commented:
Looks good. Minor comments, except for the uuid confusion @raphaelsc found. We should stop using plain uuid, they can be easily mixed up (like plain bool).

bhalevy commented:
> Looks good. Minor comments, except for the uuid confusion @raphaelsc found.

There's no mix-up.
The table's schema()->id() is synonymous with the table uuid returned by `find_uuid(ks_name, cf_name)`.

> We should stop using plain uuid, they can be easily mixed up (like plain bool).

As a follow we can define an alias type for it so it will be clear that it's a table identifier.

avikivity commented:
> > Looks good. Minor comments, except for the uuid confusion @raphaelsc found.
> 
> There's no mix-up. The table's schema()->id() is synonymous with the table uuid returned by `find_uuid(ks_name, cf_name)`. (and it's the same on all shards, for the lifetime of the table)
> 
> > We should stop using plain uuid, they can be easily mixed up (like plain bool).
> 
> As a follow we can define an alias type for it so it will be clear that it's a table identifier.

Ok. But you mean a non-alias type so the compiler complains when you pass the wrong thing.

bhalevy commented:
> > > Looks good. Minor comments, except for the uuid confusion @raphaelsc found.
> > 
> > 
> > There's no mix-up. The table's schema()->id() is synonymous with the table uuid returned by `find_uuid(ks_name, cf_name)`. (and it's the same on all shards, for the lifetime of the table)
> > > We should stop using plain uuid, they can be easily mixed up (like plain bool).
> > 
> > 
> > As a follow we can define an alias type for it so it will be clear that it's a table identifier.
> 
> Ok. But you mean a non-alias type so the compiler complains when you pass the wrong thing.

Yeah, it should probably be a struct/class that wraps a UUID with proper explicit conversion method and hash and printer implementation.

We should do the same for the planned use of UUID for sstable generation (we already have an abstraction for generation_type since d7c90b52396)

bhalevy commented on replica/table.cc:
```diff
@@ -1343,9 +1343,27 @@ future<> table::write_schema_as_cql(database& db, sstring dir) const {
+        tlogger.debug("Taking snapshot of {}.{}: directory={}", s->ks_name(), s->cf_name(), jsondir);
+
+        // FIXME: refactor per-shard snapshot and
+        // run the centralized code here
```
It's not a matter of being heavy or not.
The sharded design is problematic since it's fragile, relying on a global snapshot_manager object and aborting if it can't be allocated.
Doing the orchestration here simplifies everything and especially the error case, where we can simply return an error upstream without hanging or aborting.

denesb replied:
That is not what I meant. I'm asking about the random round-robin load balancing policy that seems to be used here. I saw later that there is quite a lot of code added here, so I guess it doesn't hurt.

bhalevy replied:
Yes, describing the schema and preparing the mainfest were already considered heavy enough to be load-balanced.
In general I think we should avoid overloading shard 0 with administrative work (and associated memory consuption), but rather we should try to distribute the load evenly across the shards.

For example, we discussed one another discussion thread having sharded<table> instances.
We shouldn't create and register those only on replica::database shard 0.
A better way would be to use a hash of the table id to select a shard responsible for managing the singleton sharded instance.

bhalevy commented:
In v2 (909fff0cc9):
- database: add truncate_table_on_all_shards
  - Moved info log message to database (with no typo)
  - and print also with/without snapshot
- database: drop_column_family: move log message up a layer
  - promote to info, and print also with/without snapshot
- database: drop_column_family: reuse uuid for evict_all_for_table
  - Added git commit message comment about the equality of schema->id and table uuid.
- database: drop_table_on_all_shards: truncate and stop table in upper layer
  - rename drop_column_family to detach_column_family
- database: drop_table_on_all_shards: use coroutine::return_exception_ptr
  - patch dropped from series
- database: add snapshot_table_on_all_shards
  - use legacy parallel_for_each instead of coroutine.
- database: truncate: move check for schema_tables keyspace to all-shards layer
  - renamed from "... system keyspace"
- utils: get rid of joinpoint
  - newly added
- table: take_snapshot: return foreign unique ptr
  - improved explanation about the motivation in the git commit message
- table: coroutinize seal_snapshot
  - extracted the call to maybe_yield() out to:
    table: seal_snapshot: maybe_yield when iterating over the table names

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1669/

bhalevy commented:
@scylladb/scylla-maint please merge sooner rather than later since the series is long and touches several layers hence more susceptible to conflicts with other PRs

denesb commented:
> @scylladb/scylla-maint please merge sooner rather than later since the series is long and touches several layers hence more susceptible to conflicts with other PRs

Waiting for @raphaelsc to approve as well.

raphaelsc commented on replica/database.cc:
```diff
@@ -1003,31 +1003,45 @@ void database::remove(const table& cf) noexcept {
+        return db.detach_column_family(*table_shards[this_shard_id()]);
     });
+    auto f = co_await coroutine::as_future(truncate_table_on_all_shards(sharded_db, table_shards, truncated_at, with_snapshot));
+    co_await smp::invoke_on_all([&] {
```
even though we moved table::stop() to happen after truncate, truncate_table_on_all_shards() makes sure compaction is temporarily disabled, so truncate won't race with any ongoing compaction. good.

raphaelsc reviewed (approved):
LGTM

denesb commented:
Conflicts, please rebase.

bhalevy commented:
> Conflicts, please rebase.

@denesb 1eaacf9c4 rebased on 527b345079

denesb commented:
> > Conflicts, please rebase.
> 
> @denesb [1eaacf9](https://github.com/scylladb/scylladb/commit/1eaacf9c4476f8ebaa5f46aff5a101dabb9347a1) rebased on [527b345](https://github.com/scylladb/scylladb/commit/527b34507910ba5606864468e81718f228257450)

Will queue once CI re-approves.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1690/

bhalevy commented:
In v3 (45ce635527)
- database: truncate: fix assert about replay_position low_mark (Fixes #11231)
- database: truncate: get optional snapshot_name from caller
- database: drop_table_on_all_shards: do not accept a truncated_at timestamp_func (Fixes #11232)
- database: add truncate_table_on_all_shards
  - keep joinpoint in storage_proxy
- database: truncate: enter per-shard table async_gate in all-shards layer
  and following patches
  - define `table_truncate_state` in class
  - and later pass it to per-shard truncate call instead of individual args
- database: get rid of timestamp_func
  - first part done earlier in "database: drop_table_on_all_shards: do not accept a truncated_at timestamp_func"
  - now pass db_clock::time_point::max()

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1706/
